src/Access/DiskAccessStorage.cpp

```diff
@@ -386,21 +379,85 @@ bool DiskAccessStorage::rebuildLists()
         if (!entity)
             continue;
 
-        const String & name = entity->getName();
-        auto type = entity->getType();
-        auto & entry = entries_by_id[id];
-        entry.id = id;
-        entry.type = type;
-        entry.name = name;
-        entry.entity = entity;
-        auto & entries_by_name = entries_by_name_and_type[static_cast<size_t>(type)];
-        entries_by_name[entry.name] = &entry;
+        all_entities.emplace_back(id, entity);
     }
 
+    setAll(all_entities);
+
     for (auto type : collections::range(AccessEntityType::MAX))
         types_of_lists_to_write.insert(type);
 
-    return true;
+    failed_to_write_lists = false; /// Try again writing lists.
+    writeLists();
+}
+
+void DiskAccessStorage::setAll(const std::vector<std::pair<UUID, AccessEntityPtr>> & all_entities)
+{
+    /// This function is similar to MemoryAccessStorage::setAll().
+    boost::container::flat_set<UUID> not_used_ids;
+    std::vector<UUID> conflicting_ids;
+
+    /// Get the list of currently used IDs. Later we will remove those of them which are not used anymore.
+    for (const auto & [id, _] : entries_by_id)
+        not_used_ids.emplace(id);
+
+    /// Get the list of conflicting IDs and update the list of currently used ones.
+    for (const auto & [id, entity] : all_entities)
+    {
+        auto it = entries_by_id.find(id);
+        if (it != entries_by_id.end())
```

Comment:
```suggestion
        if (auto it = entries_by_id.find(id);
            it != entries_by_id.end())
```
and you can do same thing below, so no need for `it2`

Reply:
done, also this part of code was moved to IAccessStorage.cpp to be used both by MemoryAccessStorage and DiskAccessStorage.